### DISCUSSION TITLE
callback in watch() is called twice with setValue()

### DISCUSSION THREAD

**Discussion by Darkspice** [react-hook-form/react-hook-form#9587] [Q&A] [answered]
Version Number
7.40.0

Why is this working like that? It works only with setValue call.
You can see a simple example in the Sandbox.

Sandbox: https://codesandbox.io/s/watch-callback-lrj0o4?file=/src/Form.tsx

**Comment by leapful** [comment, marked as answer]
@Darkspice 

By default, `setValue` will trigger watch subscription but when trying to set the target value to form field reference, there is no real field element is registered in your example therefore another call to watch subscription is triggered as fallback for this case.

**Reply by bluebill1049** [reply]
~~thanks @leapful I will give it a try to eliminate the extra render. https://github.com/react-hook-form/react-hook-form/issues/9611~~

**Reply by TimNZ** [reply]
@leapful but why?  I haven't seen an explanation for why it has to be this way.
No callback should be called twice with identical values.

**Comment by kracas** [comment]
I don't know if this is related, but watch callback is also called twice when changing a form field in Controller using setValue. However in this case the field element is real.
Sandbox: [https://codesandbox.io/s/watch-callback-forked-wfgyjz](https://codesandbox.io/s/watch-callback-forked-wfgyjz)
